### DISCUSSION TITLE
Two vscode taskfile extensions but none usable

### DISCUSSION THREAD

**Discussion by ssbarnea** [go-task/task#753] [Q&A]
There are two extensions for taskfile on marketplace but none of them is fully functioning. See https://marketplace.visualstudio.com/search?term=taskfile&target=VSCode&category=All%20categories&sortBy=Relevance

I tried both and none worked, but the funny aspect is that in order to use the JSON schema and to have auto-complete all a user needs to install is the [YAML Extension](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml) -- nothing else. In fact both the other extensions rely on that one but add few other things.

I opened this thread because it would worth having an official extension for the project, one that would be maintained by multiple people and hosted under `go-task` organization. That should address the abandonware issue with vscode extensions, which is a known plague among one-man-extensions. I also suspect that probably the two authors may be interested in this approach.

I did something similar but a bigger scale with Ansible extensions, where I "collapsed" ~10 extensions into a single official one that is now well maintained. I can probably also help with that.

**Comment by andreynering** [comment]
Hey @ssbarnea,

Yeah, the summary of the problem is that we need people to maintain it. In fact, I like that [the community has been maintaining some integrations themselves](https://taskfile.dev/community/), because I wouldn't be able to do that myself. I haven't started one because I want to focus on the tool itself primarily, which is a considerable amount of work already.

That said, if we happen to have enough people interested in maintaining that integration (being you the first one?), I am more than open to have an official repository and give the people interested write access to it. Let me know once you decide you're really interested.

VSCode is also my editor of choice, so I could at least help to test the integration 😉 

/cc @paulvarache @iulian-radu-at

**Comment by iulianraduat** [comment]
Hi @andreynering,

Unfortunately GO is not my programming language so it is hard for me to contribute.

What will be great for us (who wrote extensions for task) is to have access to a new type of output:

* `json` - contains a list of all defined tasks, including in which file is a task defined so I can jump to its definition

Now I need to parse the text output to identify which tasks are available. And I cannot jump to the file having a task definition as all I know is for which Taskfile.yml I got that task.

A possible format for this new json output (it must not need to be pretty formatted):
```json
{
"taskfiles": [
  "file": "/.../Taskfile.yml",
  "tasks": [
     "name": "inc:task-1",
     "description": "This task will do that",
     "originFile": "/.../includeMeTaskfile.yml",
     "originName": "included-as-task-1",
   ]
]
}
```

In the case of a task aliasing a task defined in an other file (even when there are multiple "redirections") it will be great to know which is the "origin" task, the one actually run.

Waiting for more feedback from the community 😃 

Cheers,
Iulian

**Reply by ssbarnea** [reply]
Clearly we need a JSON output format for `task -l`, one that would be used by other tools, like extension. Parsing its current output is dangerous and lacks info about dependencies, while parsing taskfiles themselves is also bad as you will never be able to fully reimplement what task does in javascript.

I am not sure what would be the optimal JSON format but it needs to include task names in a single mapping/dictionary, in order to keep parsing simple. Probably it could include a key that specifies from which taskfile a command originates from, like your example.

**Reply by andreynering** [reply]
An issue was created for that: #764

**Reply by ghostsquad** [reply]
How would this work without being able to evaluate variables? Given that almost anything in task can be described using templating and/or shell outputs (for variables). Seems dangerous.

**Reply by andreynering** [reply]
@ghostsquad That's a good point, but we have code written already to traverse the Taskfile without evaluating variables (or maybe sh variables only, not sure). This is used by --list, for example.

So I think we should do that here as well as the intention is to return metadata about tasks, without running anything.

**Reply by ghostsquad** [reply]
`includes` is going to be the only sticking point. You can get the keys, but the values are subject to templating.
